yksgi/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+use llvm_sys::core::*;
+use llvm_sys::{LLVMOpcode, LLVMTypeKind};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::ptr;
+
+mod llvmbridge;
+use llvmbridge::{LLVMModule, LLVMBasicBlock, LLVMValue, LLVMType, get_aot_original};
+
+/// Stopgap interpreter values.
+#[derive(Clone, Copy, PartialEq)]
+pub struct SGValue {
+    pub val: u64,
+    pub ty: LLVMType,
+}
+
+impl SGValue {
+    pub fn new(val: u64, ty: LLVMType) -> Self {
+        SGValue { val, ty }
+    }
+
+    pub fn with_type(&self, ty: LLVMType) -> Self {
+        SGValue::new(self.val, ty)
+    }
+
+    pub fn as_str(&self) -> String {
+        format!("SGValue ( {}, {:?} )", self.val, unsafe { self.ty.as_str() })
+    }
+}
+
+/// A frame holding live variables.
+struct Frame {
+    vars: HashMap<LLVMValue, SGValue>,
+}
+
+impl Frame {
+    fn new() -> Frame {
+        Frame {
+            vars: HashMap::new(),
+        }
+    }
+
+    /// Get the value of the variable `key` in this frame.
+    fn get(&self, key: &LLVMValue) -> Option<&SGValue> {
+        self.vars.get(key)
+    }
+
+    /// Insert new variable into this frame.
```

Comment:
The comment ("insert *new variable") doesn't match the code (which adds/updates a variable).

Reply:
Fixed in fcae4c6

Reply:
On reflection, I suggest moving the "add" part of this to the `init_live` function (which I *think* is the only time this functionality is needed?) and calling this function `update_var` so that it can only update an existing slot.

Reply:
It's also used in `set_var`. But I just realised that due to LLVM being SSA, we should never update a variable anyway, so this should just be insert. Let me double check this is actually the case and there wasn't a reason I added the update case.

Reply:
Fixed in 10e488a